velox/functions/lib/DateTimeFormatter.h

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/Timestamp.h"
+
+namespace facebook::velox::functions {
+
+static constexpr size_t kStructReserveSize{8};
+static constexpr size_t kLiteralReserveSize{64};
+
+enum class DateTimeFormatSpecifier : uint8_t {
+  // Era, e.g: "AD"
+  ERA = 0,
+
+  // Century of era (>=0), e.g: 20
+  CENTURY_OF_ERA = 1,
+
+  // Year of era (>=0), e.g: 1996
+  YEAR_OF_ERA = 2,
+
+  // Year of era (>=0), e.g: 1996
+  WEEK_YEAR = 3,
+
+  // Week of weekyear, e.g: 27
+  WEEK_OF_WEEK_YEAR = 4,
+
+  // Day of week, e.g: 2
+  DAY_OF_WEEK = 5,
+
+  // Day of week, e.g: "Tuesday" or "Tue", depending on number of times provided
+  // formatter character repeats
+  DAY_OF_WEEK_TEXT = 6,
+
+  // Year, e.g: 1996
+  YEAR = 7,
+
+  // Day of year, e.g: 189
+  DAY_OF_YEAR = 8,
+
+  // Month of year, e.g: "July", "Jul", 07, or 7 depending on number of times
+  // provided formatter character repeats
+  MONTH_OF_YEAR = 9,
+
+  // Day of month, e.g: 10, 01, 001, with/without padding 0s depending on number
+  // of times provided formatter character repeats
+  DAY_OF_MONTH = 10,
+
+  // Halfday of day, e.g: "PM"
+  HALFDAY_OF_DAY = 11,
+
+  // Hour of halfday (0~11)
+  HOUR_OF_HALFDAY = 12,
+
+  // Clockhour of halfday (1~12)
+  CLOCK_HOUR_OF_HALFDAY = 13,
+
+  // Hour of day (0~23)
+  HOUR_OF_DAY = 14,
+
+  // Clockhour of day (1~24)
+  CLOCK_HOUR_OF_DAY = 15,
+
+  // Minute of hour, e.g: 30
+  MINUTE_OF_HOUR = 16,
+
+  // Second of minute, e.g: 55
+  SECOND_OF_MINUTE = 17,
+
+  // Fraction of second, e.g: 978
+  FRACTION_OF_SECOND = 18,
+
+  // Timezone, e.g: "Pacific Standard Time" or "PST"
+  TIMEZONE = 19,
+
+  // Timezone offset/id, e.g: "-0800", "-08:00" or "America/Los_Angeles"
+  TIMEZONE_OFFSET_ID = 20,
+
+  // A literal % character
+  LITERAL_PERCENT = 21
+};
+
+struct FormatPattern {
+  DateTimeFormatSpecifier specifier;
+  size_t minRepresentDigits;
```

Comment:
why `minRepresentDigits`? Doesn't this represent how many times the token was consecutively repeated?

Reply:
I found the pattern to be somewhat tricky. Take year for example:
In Joda to express year 1999:
Count 1: Y -> 1999
Count 2: YY -> 99
Count 3: YYY -> 1999
Count 4: YYYY -> 1999
So we cannot simple use the repeating count to parse. I used an indirect way that will work for both joda and mysql and in general other costumed format:
Count 1: Y -> minRepresentDigits 4 -> 1999
Count 2: YY -> minRepresentDigits 2 -> 99 
Count 3: YYY -> minRepresentDigits 4 -> 1999
Count 4: YYYY -> minRepresentDigits 4 -> 1999
There is this indirection (that we'll need a couple switch or if conditions to take care of).
Similarly for mysql case it is even simpler, simple a one to one mapping:
Count 1: %Y -> minRepresentDigits 4 -> 1999

PS: I think the current implementation might not be correct in terms of above.

Reply:
I see. We should probably have a comment explaining this somewhere in the codebase.

Reply:
In case of "MONTH_OF_YEAR_TEXT", let say, what meaning "minRepresentDigits" would take?